src/components/Chat/Chatroom.js

```diff
@@ -0,0 +1,89 @@
+import React, { useEffect, useRef, useState } from "react";
+import styled from "styled-components";
+import { throttle } from "lodash";
+
+import Loading from "../shared/Loading";
+import MessageBox from "./MessageBox";
+import axios from "axios";
+import { useParams } from "react-router-dom";
+
+const ChatRoomContainer = styled.div`
+  width: 100%;
+  height: 90%;
+  overflow: auto;
+`;
+
+const TargetDiv = styled.div`
+  width: 100%;
+  height: 10px;
+`;
+
+const getMessages = async (goalId, nextPageToken) => {
+  const url = nextPageToken
+    ? `/api/chats/${goalId}?nextPageToken=${nextPageToken}`
+    : `/api/chats/${goalId}`;
+  const { data } = await axios.get(url);
+  return data.result;
+};
+
+export default function Chatroom() {
+  const scrollTarget = useRef();
+  const [target, setTarget] = useState(null);
+  const [messages, setMessages] = useState([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [token, setToken] = useState("");
+  const { id } = useParams();
+
+  const getData = async (id, token) => {
+    const prevHeight = scrollTarget.current.scrollHeight;
+
+    try {
+      const { messages, nextPageToken } = await getMessages(id, token);
+      setMessages(prev => messages.concat(prev));
+      setToken(nextPageToken);
+      scrollTarget.current.scrollTop =
+        scrollTarget.current.scrollHeight - prevHeight;
+    } catch (error) {
+      console.log(error);
+    }
+  };
+
+  const handleIntersection = async ([entry], observer) => {
+    if (entry.isIntersecting && !isLoading) {
+      observer.unobserve(entry.target);
+      setIsLoading(true);
+
+      const nextPageToken = token.length ? token : undefined;
+      await getData(id, nextPageToken);
+
+      setIsLoading(false);
+
+      if (nextPageToken) {
+        observer.observe(entry.target);
+      }
+    }
+  };
+
+  useEffect(() => {
+    let observer;
+
+    if (target) {
+      observer = new IntersectionObserver(throttle(handleIntersection, 1000), {
+        threshold: 1,
+      });
+      observer.observe(target);
+    }
+
+    return () => observer && observer.disconnect();
+  }, [target]);
+
+  return (
+    <ChatRoomContainer ref={scrollTarget}>
+      {isLoading && <Loading />}
+      {<TargetDiv ref={setTarget}></TargetDiv>}
+      {messages.map((data, i) => {
+        return <MessageBox key={i} data={data} />;
+      })}
```

Comment:
```suggestion
import {v4 as uuidv4} from "uuid";

      {messages.map(data => {
        return <MessageBox key={uuidv4()} data={data} />;
      })}
```
현재는 항목의 순서가 바뀔 수 있는 경우는 아니지만, key값으로 index를 주는 것은 좋지 않다고 [공식문서](https://ko.reactjs.org/docs/lists-and-keys.html)에 나와 있더라구요!
key 값으로 설정하기에 마땅한 id 값이 없다면 uuid()를 사용하여 고유 값을 줄 수도 있을 것 같습니다👍
현재 데이터 값으로는 키 값으로 줄만한 값이 없어서 uuid를 사용해보는 것도 좋을 것 같아요!

Reply:
목데이터로 테스트하다 보니 i를 쓰게 되었는데, 실제 데이터에서는 메시지 생성 시간을 활용해도 괜찮을 것 같아요!